src/Toml/Generic.hs

```diff
@@ -0,0 +1,390 @@
+    genericTomlCodec options = (:*:)
+        <$> genericTomlCodec options .= fstG
+        <*> genericTomlCodec options .= sndG
+      where
+        fstG :: (f :*: g) p -> f p
+        fstG (f :*: _) = f
+
+        sndG :: (f :*: g) p -> g p
+        sndG (_ :*: g) = g
+    {-# INLINE genericTomlCodec #-}
+
+instance (Selector s, HasCodec a) => GenericCodec (S1 s (Rec0 a)) where
+    genericTomlCodec GenericOptions{..} = genericWrap $ hasCodec @a fieldName
+      where
+        genericWrap :: TomlCodec a -> TomlCodec (S1 s (Rec0 a) p)
+        genericWrap = Toml.dimap (unK1 . unM1) (M1 . K1)
+
+        fieldName :: Key
+        fieldName =
+            fromString
+            $ genericOptionsFieldModifier
+            $ selName (error "S1" :: S1 s Proxy ())
+    {-# INLINE genericTomlCodec #-}
+
+----------------------------------------------------------------------------
+-- Helper typeclasses
+----------------------------------------------------------------------------
+
+{- | This typeclass tells how the data type should be coded as an item of a
+list. Lists in TOML can have two types: __primitive__ and __table of arrays__.
+
+* If 'hasItemCodec' returns 'Left': __primitive__ arrays codec is used.
+* If 'hasItemCodec' returns 'Right:' __table of arrays__ codec is used.
+-}
+class HasItemCodec a where
+    hasItemCodec :: Either (TomlBiMap a AnyValue) (TomlCodec a)
+
+instance HasItemCodec Bool      where hasItemCodec = Left Toml._Bool
+instance HasItemCodec Int       where hasItemCodec = Left Toml._Int
+instance HasItemCodec Word      where hasItemCodec = Left Toml._Word
+instance HasItemCodec Integer   where hasItemCodec = Left Toml._Integer
+instance HasItemCodec Natural   where hasItemCodec = Left Toml._Natural
+instance HasItemCodec Double    where hasItemCodec = Left Toml._Double
+instance HasItemCodec Float     where hasItemCodec = Left Toml._Float
+instance HasItemCodec Text      where hasItemCodec = Left Toml._Text
+instance HasItemCodec L.Text    where hasItemCodec = Left Toml._LText
+instance HasItemCodec ZonedTime where hasItemCodec = Left Toml._ZonedTime
+instance HasItemCodec LocalTime where hasItemCodec = Left Toml._LocalTime
+instance HasItemCodec Day       where hasItemCodec = Left Toml._Day
+instance HasItemCodec TimeOfDay where hasItemCodec = Left Toml._TimeOfDay
+instance HasItemCodec IntSet    where hasItemCodec = Left Toml._IntSet
+
+{- | If data type @a@ is not primitive then this instance returns codec for list
+under key equal to @a@ type name.
+-}
+instance (HasItemCodec a, Typeable a) => HasItemCodec [a] where
+    hasItemCodec = case hasItemCodec @a of
+        Left prim   -> Left $ Toml._Array prim
+        Right codec -> Right $ Toml.list codec (fromString $ typeName @a)
+
+{- | Helper typeclass for generic deriving. This instance tells how the data
+type should be coded if it's a field of other data type.
+
+__NOTE:__ You can use this typeclass for writing your custom codecs manually but
```

Comment:
You mean, you can write instances for any custom data types, but this is not recommended way? The alternative is using one of the generic codecs?

Reply:
What I meant actually, is that instead of writing:

```haskell
userCodec :: TomlCodec User
userCodec = User
    <$> Toml.text "name" .= userName
    <*> Toml.int  "age"  .= userAge
```

You can write

```
userCodec :: TomlCodec User
userCodec = User
    <$> Toml.hasCodec "name" .= userName
    <*> Toml.hasCodec "age"  .= userAge
```

But it's not recommended. Probably should write this in a more clear way :slightly_smiling_face:

Reply:
Updated docs for this typeclass.